rfcs/2021-04-07-6330-emitting-multiple-log-events-from-remap-transform.md

```diff
@@ -0,0 +1,112 @@
+
+## Doc-level Proposal
+
+The `remap` tranform would gain an extra configuration option:
+
+```text
+emit_root = true/false # default false
+```
+
+When `emit_root` is `true`, the value of `.` will be emitted at the end of the remap program. When `emit_root` is false, the value of `.` will not be emitted. Instead users should use the `emit_log` function to emit.
+
+## Rationale
+
+This enhances remap to support additional transformation use-cases of splitting up an event into multiple events. Without this, users will continue to have to use Lua to acheive this, which introduces a performance bottleneck compared to remap
+
+## Prior Art
+
+- [fluent-plugin-record_splitter](https://github.com/ixixi/fluent-plugin-record_splitter)
+- [Logstash split](https://www.elastic.co/guide/en/logstash/current/plugins-filters-split.html)
+
+These are similar to the `explode` transform PR we have: https://github.com/timberio/vector/pull/6545
+
+## Drawbacks
+
+- It makes the remap transform more complicated to understand and adds an additional configuration option that users need to be aware of
+- Ongoing maintenance burden should be minimal
+
+## Alternatives
+
+### Not emitting . if an `emit_log` is present in the program
+
+We could avoid having an `emit_root` config option on the remap transform by just not emitting automatically if we see an `emit_log` function in the user-provided source. I personally think this would be a bit suprising, but it is an option.
+
+### Modifying remap to accept setting the root object to an array
+
+https://github.com/timberio/vector/issues/6988
+
+This would modify remap to allow setting `.` to an array of objects to have each element emitted independently.
+
+This turned out to require a bigger change than I expected in that `.` is linked to mutating the underlying event (metric or log). It's definitely doable, but would require a substantial refactoring and so caused me to take a step back and consider the alternatives, prompting this RFC.
+
+I also don't see an easy way to extend this to deal with emitting multiple metrics without requiring users to create objects with a known format (like `{ "name": "foo", "namespace": "bar", "type = "counter", "value": 1 ...}`) whereas it seems more straightforward to extend the proposed approach with functions like `emit_counter`
+
+### An `explode` transformation
+
+https://github.com/timberio/vector/pull/6545
+
+Have a separate, special purpose, transform for converting a single log event into multiple log events.
+
+This would work, but I think the downsides are:
+
+- Introduces a new transform rather than just leveraging an existing one, remap. There is some precedence for this with the `route` and `reduce` transforms though.
+- It is less flexible than letting users emit arbitrary events
+- There isn't a clear path to emitting, say, multiple metrics from an input log event
```

Comment:
If we _don't_ want to introduce a new transform, then I'd probably lean towards updating the `remap` transform with a new option `explode = true`. If set, the transform would check at boot that your VRL script returns an array, and it would behave the same as what I described above, except that it doesn't require a separate transform and allows you to mutate the event before you explode it into zero or more events.

I still lean towards a separate transform being the best option, but this would probably be my second preference.

I realize that both of my proposals are _less flexible_ than an `emit_log`, but I think overall they come out on top due to their simplicity over the current proposal in the RFC.

Reply:
Thanks for the alternate proposal!

I like the idea of a separate transform where it evaluates the last expression (similar to conditions) to get the output. This should avoid the need to decouple `.` from the underlying event for now.

> I still somewhat prefer a separate transform for this. I worry that — in a couple of months — we'll be at a point where the remap transform is a "do anything you want" transform, which it was never meant to be.

Agreed, I can see that this is expanding the scope significantly. I think I do see a future where remap is more general purpose, similar to Lua, to offer increased flexibility, but we can defer that for now until we have more data from users about what they are trying to achieve.

> > The biggest open question for me with this approach is whether the event in this transform should be _immutable_, to enforce people to use `remap` to mutate events, and `explode` to convert to multiple events, instead of using `explode` to do both.
>
> I'd probably lean to **yes**, make the event immutable, if nothing else, it gives us more flexibility down the line on what we want to do with this transform, and it also guides operators to use our transforms "for their intended purpose". We can always remove that restriction in the future, if it turns out to be a good idea to do so.

I also lean towards yes, it should be immutable in that modifications should not persist. I do think it should be possible to do some mapping in the `explode` transform though. For example, I think it will be common to want to map over the elements that are being exploded to add additional fields like:

Input:

```
{ "host": "foobar", "events": [{ "message": "foo" }, { "message": "bar" }] }
```

```toml
[transforms.explode]
type = "explode"
source = "map(array!(.events), |event| event.host = .host) ?? []"
```

The exact "mapping over elements" syntax will be contingent on #6031 .

I'll update the RFC to reflect this approach as I think it is a simpler start and we can always add in `emit_log` later if we find an eed.